reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt

```diff
@@ -2,41 +2,32 @@ package com.badoo.reaktive.single
 
 import com.badoo.reaktive.disposable.Disposable
 import com.badoo.reaktive.disposable.DisposableWrapper
+import com.badoo.reaktive.disposable.doIfNotDisposed
 
 fun <T> single(onSubscribe: (emitter: SingleEmitter<T>) -> Unit): Single<T> =
     singleUnsafe { observer ->
-        val disposableWrapper = DisposableWrapper()
-        observer.onSubscribe(disposableWrapper)
-
         val emitter =
-            object : SingleEmitter<T> {
-                override val isDisposed: Boolean get() = disposableWrapper.isDisposed
-
+            object : DisposableWrapper(), SingleEmitter<T> {
                 override fun onSuccess(value: T) {
-                    if (!disposableWrapper.isDisposed) {
-                        try {
-                            observer.onSuccess(value)
-                        } finally {
-                            disposableWrapper.dispose()
-                        }
+                    doIfNotDisposed(dispose = true) {
+                        observer.onSuccess(value)
                     }
                 }
 
                 override fun onError(error: Throwable) {
-                    if (!disposableWrapper.isDisposed) {
-                        try {
-                            observer.onError(error)
-                        } finally {
-                            disposableWrapper.dispose()
-                        }
+                    doIfNotDisposed(dispose = true) {
+                        observer.onError(error)
                     }
                 }
 
+                @Suppress("RedundantOverride") // IDEA complains that setDisposable is not implemented
```

Comment:
Only here and not in completable, maybe and observable?

Reply:
Mistake, will fix